gssapi/raw/ext_krb5.pyx

```diff
@@ -0,0 +1,577 @@
+
+    cdef struct gss_krb5_rfc1964_keydata:
+        OM_uint32 sign_alg
+        OM_uint32 seal_alg
+        gss_krb5_lucid_key_t ctx_key
+    ctypedef gss_krb5_rfc1964_keydata gss_krb5_rfc1964_keydata_t
+
+    cdef struct gss_krb5_cfx_keydata:
+        OM_uint32 have_acceptor_subkey
+        gss_krb5_lucid_key_t ctx_key
+        gss_krb5_lucid_key_t acceptor_subkey
+    ctypedef gss_krb5_cfx_keydata gss_krb5_cfx_keydata_t
+
+    cdef struct gss_krb5_lucid_context_v1:
+        OM_uint32 version
+        OM_uint32 initiate
+        OM_uint32 endtime
+        uint64_t send_seq
+        uint64_t recv_seq
+        OM_uint32 protocol
+        gss_krb5_rfc1964_keydata_t rfc1964_kd
+        gss_krb5_cfx_keydata_t cfx_kd
+    ctypedef gss_krb5_lucid_context_v1 gss_krb5_lucid_context_v1_t
+
+    gss_OID GSS_KRB5_NT_PRINCIPAL_NAME
+    int32_t _PY_GSSAPI_KRB5_TIMESTAMP
+
+    # The krb5 specific types are defined generically as the type names differ
+    # across GSSAPI implementations.
+
+    OM_uint32 gss_krb5_ccache_name(OM_uint32 *minor_status, const char *name,
+                                   const char **out_name) nogil
+
+    OM_uint32 gss_krb5_export_lucid_sec_context(OM_uint32 *minor_status,
+                                                gss_ctx_id_t *context_handle,
+                                                OM_uint32 version,
+                                                void **kctx) nogil
+
+    # The actual authtime size differs across implementations, see method for
+    # more info.
+    OM_uint32 gsskrb5_extract_authtime_from_sec_context(
+        OM_uint32 *minor_status, gss_ctx_id_t context_handle,
+        void *authtime) nogil
+
+    OM_uint32 gsskrb5_extract_authz_data_from_sec_context(
+        OM_uint32 *minor_status, const gss_ctx_id_t context_handle,
+        int ad_type, gss_buffer_t ad_data) nogil
+
+    OM_uint32 gss_krb5_free_lucid_sec_context(OM_uint32 *minor_status,
+                                              void *kctx) nogil
+
+    OM_uint32 gss_krb5_import_cred(OM_uint32 *minor_status,
+                                   void *id,  # krb5_ccache
+                                   void *keytab_principal,  # krb5_principal
+                                   void *keytab,  # krb5_keytab
+                                   gss_cred_id_t *cred) nogil
+
+    # MIT uses a int32_t whereas Heimdal uses uint32_t. Use void * to satisfy
+    # the compiler.
+    OM_uint32 gss_krb5_get_tkt_flags(OM_uint32 *minor_status,
+                                     gss_ctx_id_t context_handle,
+                                     void *ticket_flags) nogil
+
+    OM_uint32 gss_krb5_set_allowable_enctypes(OM_uint32 *minor_status,
+                                              gss_cred_id_t cred,
+                                              OM_uint32 num_ktypes,
+                                              int32_t *ktypes) nogil
+
+
+cdef class Krb5LucidContext:
+    # cdef void *raw_ctx
```

Comment:
The context needs to be stored somewhere with a reference and freed/deallocated by Python once all references are dropped. The reason why I went with a class instead of having `krb5_export_lucid_sec_context` return a named tuple with the values already copied is:

* The function takes in a version so in the future different contexts could be returned
* This is the base class that stores the raw lucid pointer and the `__dealloc__` function
* A version x specific class inherits this and implements the properties to retrieve the version specific properties
* In the future there could be a version 2 that also inherits this class to gain the free logic  without copying the code again

Technically I could have `krb5_export_lucid_sec_context` return a named tuple for version 1 and have that free the lucid context pointer in the function but I'm unsure whether to hard code a failure if version isn't specified.

Reply:
Apologies, my question wasn't clear.  What I meant to ask about was specifically the comment.

Reply:
That seems to be the standard for this libary. The attributes are defined in the `.pxd` and a comment is added. I guess this is to just document the code where the attribute is actually used that it's there.

See https://github.com/pythongssapi/python-gssapi/blob/main/gssapi/raw/sec_contexts.pxd#L3-L5 and https://github.com/pythongssapi/python-gssapi/blob/main/gssapi/raw/sec_contexts.pyx#L77-L82. I should add in the line `# defined in pxd` though to make it clearer.